### PR TITLE
uploader: fix HTTP status code handling for redirects

### PR DESCRIPTION
**Pull request by lineman60**
# Pull Request Check List

Resolves: #7159

- [x] Added **tests** for changed code.
- [N/A ] Updated **documentation** for changed code.

### CODE REVIEW

File: src/poetry/publishing/uploader.py [hunk 1]
```diff
@@ -261,12 +261,12 @@ def _upload_file(
                         headers={"Content-Type": monitor.content_type},
                         timeout=REQUESTS_TIMEOUT,
                     )
-                if resp is None or 200 <= resp.status_code < 300:
+                if resp is None or 200 <= resp.status_code < 299:
```
Comment (dimbleby) [inline src/poetry/publishing/uploader.py:264]: you have introduced an out by one error here and elsewhere (what happens for 299? etc)

File: src/poetry/publishing/uploader.py [hunk 2]
```diff
@@ -275,7 +275,9 @@ def _upload_file(
                         "Redirects are not supported. "
                         "Is the URL missing a trailing slash?"
                     )
-                elif resp.status_code == 400 and "was ever registered" in resp.text:
+                elif (
```
Comment (dimbleby) [inline src/poetry/publishing/uploader.py:278]: this code branch looks as though it was aimed at the specific behaviour of a particular server (almost surely pypi).  I doubt that broadening the range of status codes is useful

File: src/poetry/publishing/uploader.py [hunk 3]
```diff
@@ -261,12 +261,12 @@ def _upload_file(
...
                         timeout=REQUESTS_TIMEOUT,
                     )
-                if resp is None or 200 <= resp.status_code < 300:
+                if resp is None or 200 <= resp.status_code < 299:
                     bar.set_format(
                         f" - Uploading <c1>{file.name}</c1> <fg=green>%percent%%</>"
                     )
                     bar.finish()
-                elif resp.status_code == 301:
+                elif 300 <= resp.status_code < 399:
```
Comment (hutson) [inline src/poetry/publishing/uploader.py:269]:
```suggestion
                elif 300 <= resp.status_code < 400:
```

File: src/poetry/publishing/uploader.py (hunk 2 above)
Comment (hutson) [inline src/poetry/publishing/uploader.py:278]: Seems unrelated to addressing the behavior around 3xx codes - https://github.com/python-poetry/poetry/issues/7159

Comment (lineman60) [reply src/poetry/publishing/uploader.py]: Removed this change.

File: src/poetry/publishing/uploader.py (hunk 1 above)
Comment (lineman60) [reply src/poetry/publishing/uploader.py]: Thanks, fixed!

File: src/poetry/publishing/uploader.py (hunk 3 above)
Comment (lineman60) [reply src/poetry/publishing/uploader.py]: updated to 400

**Comment by dimbleby** [comment]
This looks harmless to me, but I don't follow why this fix is necessary.  

Wouldn't the code previously have gone through the `else ... resp.raise_for_status()` branch?  and then through the `except ... requests.HTTPError` branch, raising an `UploadError`?

ie the command should have failed all along, just without the text about "Redirects are not supported..."?

File: tests/publishing/test_uploader.py [hunk 4]
```diff
@@ -65,6 +65,24 @@ def test_uploader_properly_handles_nonstandard_errors(
...
+        (303, "See Other"),
+        (302, "Found"),
+        (301, "Moved Permanently"),
+        (300, "Multiple Choices"),
+    ],
+)
+def test_uploader_properly_handles_redirects(
+    http: type[httpretty.httpretty], uploader: Uploader, status: int, body: str
+):
+    http.register_uri(http.POST, "https://foo.com", status=status, body=body)
```
Comment (radoering) [inline tests/publishing/test_uploader.py:83]: Is something missing here? The test succeeds even without the change.

Comment (lineman60) [reply tests/publishing/test_uploader.py:83]: @radoering  I for got the assert, should be fixed now.

**Review by radoering — APPROVED** [review]
